clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -701,3 +808,95 @@ void CGHLSLRuntime::emitInitListOpaqueValues(CodeGenFunction &CGF,
     }
   }
 }
+
+std::optional<LValue> CGHLSLRuntime::emitResourceArraySubscriptExpr(
+    const ArraySubscriptExpr *ArraySubsExpr, CodeGenFunction &CGF) {
+  assert(ArraySubsExpr->getType()->isHLSLResourceRecord() ||
+         ArraySubsExpr->getType()->isHLSLResourceRecordArray() &&
+             "expected resource array subscript expression");
+
+  // let clang codegen handle local resource array subscrips
+  const VarDecl *ArrayDecl = dyn_cast<VarDecl>(getArrayDecl(ArraySubsExpr));
+  if (!ArrayDecl || !ArrayDecl->hasGlobalStorage())
+    return std::nullopt;
+
+  // FIXME: this is not yet implemented (llvm/llvm-project#145426)
+  assert(!ArraySubsExpr->getType()->isArrayType() &&
+         "indexing of array subsets it not supported yet");
+
+  // get total array size (= range size)
+  const Type *ResArrayTy = ArrayDecl->getType().getTypePtr();
+  assert(ResArrayTy->isHLSLResourceRecordArray() &&
```

Comment:
Isn't it possible that this could be false if only the first condition of 814 is true?
I don't think getTypePtr() gets the element type, it should just strip the array type and make it canonical.
In what situation would `isHLSLResourceRecord` exit this function before we get to this line.

Reply:
The asserts at the top of this function are checking the type of the `ArraySubscriptExpr`, which is the result type of the subscript expressions. It must be a resource, or a resource array when it is indexing a subset of a multi-dimensional array).
The assert on line 829 is to make sure that whatever declaration we are indexing into is an array of resources. So it is slightly different.

Reply:
Ah, I notice that you'd probably exit on line 821 if only the first is true and the second is false on 814 anyways. 
So you'd never get false on this assert. I think it makes sense!